src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+/// <reference path="../refactorProvider.ts" />
+/// <reference path="../../compiler/checker.ts" />
+
+/* @internal */
+namespace ts.refactor.extractMethod {
+    const extractMethod: Refactor = {
+        name: "Extract Method",
+        description: Diagnostics.Extract_function.message,
+        getAvailableActions,
+        getEditsForAction,
+    };
+
+    registerRefactor(extractMethod);
+
+    /** Compute the associated code actions */
+    function getAvailableActions(context: RefactorContext): ApplicableRefactorInfo[] | undefined {
+        const rangeToExtract = getRangeToExtract(context.file, { start: context.startPosition, length: context.endPosition - context.startPosition });
+
+        const targetRange: TargetRange = rangeToExtract.targetRange;
+        if (targetRange === undefined) {
+            return undefined;
+        }
+
+        const extractions = extractRange(targetRange, context);
+        if (extractions === undefined) {
+            // No extractions possible
+            return undefined;
+        }
+
+        const actions: RefactorActionInfo[] = [];
+        const usedNames: Map<boolean> = createMap();
+
+        let i = 0;
+        for (const extr of extractions) {
+            // Skip these since we don't have a way to report errors yet
+            if (extr.errors && extr.errors.length) {
+                continue;
+            }
+
+            // Don't issue refactorings with duplicated names
+            const description = formatStringFromArgs(Diagnostics.Extract_function_into_0.message, [extr.scopeDescription]);
+            if (!usedNames.get(description)) {
+                usedNames.set(description, true);
+                actions.push({
+                    description,
+                    name: `scope_${i}`
+                });
+            }
+            i++;

```

Comment:
Looks like it's important to leave gaps in the sequence.  Consider adding a comment to that effect.

---
In reply to: [132311523](https://github.com/Microsoft/TypeScript/pull/17625#discussion_r132311523) [](ancestors = 132311523)